### PR TITLE
[WIP] Faster `List<T>` element write

### PR DESCRIPTION
**Pull request by VSadov**

### CODE REVIEW

**Comment by benaadams** [comment]
To get around array covariance checks?

**Comment by VSadov** [comment]
Yes, avoiding dynamic typechecks because of array covariance, which we do not use in List.

**Comment by benaadams** [comment]
> Yes, avoiding dynamic typechecks because of array covariance, which we do not use in List.

Other candidates in corefx would be [`Queue<T>`](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/Queue.cs#L25-L29), [`SortedList<TKey, TValue>`](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/SortedList.cs#L54-L58), and [`Stack<T>`](https://github.com/dotnet/corefx/blob/master/src/System.Collections/src/System/Collections/Generic/Stack.cs#L26-L30)

**Comment by VSadov** [comment]
`AsRef(in)`  Everything!!!   ;-)

I did not have a chance to measure the benefits though...
I’d guess there are definitely cases where this is profitable, just need to be sure no regressions in other cases.

**Comment by jkotas** [comment]
I do not think we should be spraying tricky type unsafe code everywhere to squeeze the bit of performance overhead caused by covariance checks.

If this has measurable benefits, I think we should introduce a helper type like what is proposed in https://github.com/dotnet/corefx/issues/23689 and use that instead.

**Comment by VSadov** [comment]
InvariantArray type would be nice. 
Statically inferring that array covariance is not used seems even nicer. - Array covariance is extremely rare  nowdays after generics were introduced.

However, these features may take months to implement and `AsRef(in)` is here already and does not prevent switching to something better in the future.

I do not really recommend to put it everywhere (that was obviously a joke), but for very common types like `List<T>` any perf improvement will be amplified to countless uses.
I do not see why so much resistance to this change.

**Comment by benaadams** [comment]
I assume `CORINFO_HELP_ARRADDR_ST` is more expensive than `CORINFO_HELP_CHECKED_ASSIGN_REF`?

**Comment by xoofx** [comment]
> I assume CORINFO_HELP_ARRADDR_ST is more expensive than CORINFO_HELP_CHECKED_ASSIGN_REF?

It looks like.

`CORINFO_HELP_ARRADDR_ST` is going to [translated](https://github.com/dotnet/coreclr/blob/430331c011a0460670e2743ab08caa36f914f52d/src/inc/jithelpers.h#L108-L111) to:

https://github.com/dotnet/coreclr/blob/59714b683f40fac869050ca08acc5503e84dc776/src/vm/amd64/jithelpers_fast.S#L394-L446

while `CORINFO_HELP_CHECKED_ASSIGN_REF` is [translated](https://github.com/dotnet/coreclr/blob/430331c011a0460670e2743ab08caa36f914f52d/src/inc/jithelpers.h#L171) to a simple write barrier:

https://github.com/dotnet/coreclr/blob/59714b683f40fac869050ca08acc5503e84dc776/src/vm/amd64/jithelpers_fast.S#L28-L50

going to run some benchmarks just to check

**Comment by xoofx** [comment]
I'm also interested in this new Unsafe API `public static ref T AsRef<T>(in T source)` just alone 😃

**Comment by xoofx** [comment]
Got the following results on a custom List (not the coreclr list but a copy from it which is using a struct instead of a class and removing version):

```
          Method |     Mean |     Error |    StdDev |
---------------- |---------:|----------:|----------:|
     StandardAdd | 16.57 ns | 0.0282 ns | 0.0264 ns |
 NoCovarianceAdd | 15.76 ns | 0.0575 ns | 0.0538 ns |
```

It looks like 5% improvement but should better double check with the real coreclr `List<T>`

**Comment by VSadov** [comment]
`CORINFO_HELP_CHECKED_ASSIGN_REF` should at some point perform dynamic check that assigned value is type-compatible with the actual array element type.

That could have some costs, especially if assigned values do not have the same exact type as the T. 
That could happen if T is the base type or an interface.

Example:
```C#
        static void Main(string[] args)
        {
            var l = new List<IEnumerable<object>>();            
            l.Add(l);

            for (int i = 0; i < 1000000000; i++)
            {
                // this would be mostly type checks
                l[0] = l;
            }
        }
```

**Comment by jkotas** [comment]
> However, these features may take months to implement and AsRef(in) is here already and does not prevent switching to something better in the future.

I would much rather take months to implement this than open flood gates for this.

Once it is used in one place, there is always argument that it should be used in more places. The result is that we are wasting time dealing with influx of changes that are making the code hard to maintain. The `ThrowHelper` is a great example of this dynamic.

> `CORINFO_HELP_CHECKED_ASSIGN_REF` should at some point perform dynamic check that assigned value is type-compatible with the actual array element type.

This does not make sense.

File: src/mscorlib/shared/Internal/Runtime/CompilerServices/Unsafe.cs [hunk 1]
```diff
@@ -357,6 +357,20 @@ public static void Write<T>(ref byte destination, T value)
             return ref Unsafe.As<byte, T>(ref *(byte*)source);
         }
 
+        /// <summary>
+        /// Reinterprets the given location as a reference to a value of type <typeparamref name="T"/>.
+        /// </summary>
+        [Intrinsic]
+        [NonVersionable]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ref T AsRef<T>(in T source)
```
Comment (jkotas) [inline src/mscorlib/shared/Internal/Runtime/CompilerServices/Unsafe.cs]: This works by a pure luck. The method may or may not get the overridden IL on the VM side, depending on how it is ordered with other AsRef.

You would need to disambiguate this with the other AsRef in the VM, and supply the IL for this.

Comment (VSadov) [reply src/mscorlib/shared/Internal/Runtime/CompilerServices/Unsafe.cs]: Somehow it works, but I have to agree here - if there is a way to do it correctly, it should be done correctly.

**Comment by benaadams** [comment]
> This does not make sense.

Assume meant `CORINFO_HELP_ARRADDR_ST`

**Comment by VSadov** [comment]
Ah yes - `CORINFO_HELP_ARRADDR_ST`  is the one that typechecks.

We must make a rule that more expensive helpers have longer names to avoid confusions like this. :-)

**Comment by benaadams** [comment]
Looking at variants
```csharp
ref T element = ref array[size];
element = item;
```
Causes the following changes
```diff
 G_M27982_IG01:
        4883EC28             sub      rsp, 40
-       4889542438           mov      gword ptr [rsp+38H], rdx
+       56                   push     rsi
+       4883EC20             sub      rsp, 32
+       488BF2               mov      rsi, rdx
 
 G_M27982_IG02:
        488B4908             mov      rcx, gword ptr [rcx+8]
        FF411C               inc      dword ptr [rcx+28]
        4C8B4108             mov      r8, gword ptr [rcx+8]
        8B4118               mov      eax, dword ptr [rcx+24]
-       41394008             cmp      dword ptr [r8+8], eax
+       44394208             cmp      dword ptr [rdx+8], r8d
        7619                 jbe      SHORT G_M27982_IG03
-       448D4801             lea      r9d, [rax+1]
+       418D4001             lea      eax, [r8+1]
-       44894918             mov      dword ptr [rcx+24], r9d
+       894118               mov      dword ptr [rcx+24], eax
-       498BC8               mov      rcx, r8
-       8BD0                 mov      edx, eax
+       488BCA               mov      rcx, rdx
+       418BD0               mov      edx, r8d
-       4C8B442438           mov      r8, gword ptr [rsp+38H]
-       E8ACBCA95F           call     CORINFO_HELP_ARRADDR_ST
+       49B8585798ACFA7F0000 mov      r8, 0x7FFAAC985758
+       E8A749A45F           call     CORINFO_HELP_LDELEMA_REF
+       488BC8               mov      rcx, rax
+       488BD6               mov      rdx, rsi
+       E82CBBA75F           call     CORINFO_HELP_CHECKED_ASSIGN_REF
        EB0A                 jmp      SHORT G_M27982_IG04
 
 G_M27982_IG03:
-       488B542438           mov      rdx, gword ptr [rsp+38H]
+       488BD6               mov      rdx, rsi
        E8E0EEFEFF           call     List`1:AddWithResize(ref):this
 
 G_M27982_IG04:
        90                   nop      
 
 G_M27982_IG05:
-       4883C428             add      rsp, 40
+       4883C420             add      rsp, 32
+       5E                   pop      rsi
        C3                   ret      
```
and
```csharp
(ref array[size]) = item;
```
Doesn't compile

**Comment by VSadov** [comment]
@benaadams  you have to get a _readonly_ ref to an array element - to avoid the typecheck, and then use `Unsafe.AsRef` to cast-away readonliness. 
It is only ok because we know that `List<T>` never uses the underlying array covariantly. 

In general - If you find a way to avoid the typecheck without unsafe code, it is a bug.

**Comment by benaadams** [comment]
Could add
```csharp
internal static class ArrayExtensions
{
    public static ref T RefAt<T>(this T[] array, int index)
    {
        // Workaround until https://github.com/dotnet/corefx/issues/23689
        return ref Unsafe.AsRef(in array[index]);
    }
}
```
Then changes are
```diff
-    array[size] = item;
+    array.RefAt(size) = item;
```
Rather than
```diff
-    array[size] = item;
+    Unsafe.AsRef(in array[size]) = item;
```
Both generating same asm improvements, but it moves the `Unsafe` out of flow?

**Comment by benaadams** [comment]
Could also use it for `ItemRef` https://github.com/dotnet/corefx/issues/25189
```csharp
public partial class List<T>
{
    private ref T ItemRef(int index)
    {
        T[] array = _items;
        if ((uint)index >= (uint)_size)
        {
            ThrowHelper.ThrowArgumentOutOfRange_IndexException();
        }

        return ref array.RefAt(index);
    }
}
```

**Comment by VSadov** [comment]
@benaadams `RefAt` looks interesting. I'd add a "no-covariance" assert.

```C#
internal static class ArrayExtensions
{
    public static ref T RefAt<T>(this T[] array, int index)
    {
        // Workaround until https://github.com/dotnet/corefx/issues/23689
        Debug.Assert(typeof(T[]) == array.GetType(), "wow covariance!");
        return ref Unsafe.AsRef(in array[index]);
    }
}
```

**Comment by mjp41** [comment]
@VSadov it looks to me like the Unsafe.AsRef(in) could be used to implement the InvariantArray proposal.

**Comment by benaadams** [comment]
Using `InvariantArray<T>` for `List<T>` has a crossgen issue https://github.com/dotnet/coreclr/compare/master...benaadams:ListElemWrite as it won't inline `InvariantArray``1:get_Item(int)` due to `runtime dictionary lookup` (also may cause binary serialization issues for derived classes for `List<T>` specifically, if anyone really does derive it and use binary serilaization?)

```
Inlines into 06004335 List`1:Add(ref):this
 [aggressive inline attribute] InvariantArray`1:get_Length():int:this
 [FAILED: runtime dictionary lookup] InvariantArray`1:get_Item(int):byref:this
 [FAILED: noinline per IL/cached result] List`1:AddWithResize(ref):this
```
So `Add` ends up being
```asm
Successfully inlined InvariantArray`1:get_Length():int:this (9 IL bytes) (depth 1) [aggressive inline attribute]
**************** Inline Tree
Inlines into 06004335 List`1:Add(ref):this
  [1 IL=0031 TR=000023 06004301] [aggressive inline attribute] InvariantArray`1:get_Length():int:this
  [0 IL=0050 TR=000082 06004300] [FAILED: runtime dictionary lookup] InvariantArray`1:get_Item(int):byref:this
  [0 IL=0064 TR=000066 06004336] [FAILED: noinline per IL/cached result] List`1:AddWithResize(ref):this
Budget: initialTime=270, finalTime=274, initialBudget=2700, currentBudget=2704
Budget: increased by 4 because of force inlines
Budget: initialSize=1727, finalSize=1727
; Assembly listing for method List`1:Add(ref):this
; Emitting BLENDED_CODE for X64 CPU with SSE2
; optimized code
; rsp based frame
; partially interruptible
; Final local variable assignments
;
;  V00 this         [V00,T00] ( 14, 12   )     ref  ->  rsi         this class-hnd
;  V01 arg1         [V01,T03] (  4,  3   )     ref  ->  rdi         class-hnd
;  V02 loc0         [V02    ] (  3,  2.50)  struct ( 8) [rsp+0x30]   do-not-enreg[XS] must-init addr-exposed ld-addr-op
;  V03 loc1         [V03,T04] (  6,  4.50)     int  ->  rbx        
;  V04 tmp0         [V04,T01] (  3,  6   )    long  ->  rcx        
;  V05 tmp1         [V05,T06] (  1,  2   )    long  ->  [rsp+0x28]  
;  V06 tmp2         [V06,T07] (  2,  1.50)    long  ->  rdx        
;  V07 tmp3         [V07,T05] (  5,  4.50)    long  ->  rax        
;  V08 tmp4         [V08    ] (  3,  2.50)     ref  ->  [rsp+0x30]   do-not-enreg[X] addr-exposed V02._inner(offs=0x00) P-DEP
;  V09 OutArgs      [V09    ] (  1,  1   )  lclBlk (32) [rsp+0x00]  
;  V10 cse0         [V10,T02] (  5,  4   )    long  ->  rcx        
;
; Lcl frame size = 64

G_M43649_IG01:
       57                   push     rdi
       56                   push     rsi
       53                   push     rbx
       4883EC40             sub      rsp, 64
       33C0                 xor      rax, rax
       4889442430           mov      qword ptr [rsp+30H], rax
       48894C2438           mov      qword ptr [rsp+38H], rcx
       488BF1               mov      rsi, rcx
       488BFA               mov      rdi, rdx

G_M43649_IG02:
       FF4614               inc      dword ptr [rsi+20]
       488B4E18             mov      rcx, gword ptr [rsi+24]
       48894C2430           mov      gword ptr [rsp+30H], rcx
       8B5E10               mov      ebx, dword ptr [rsi+16]
       488B0E               mov      rcx, qword ptr [rsi]
       488B4930             mov      rcx, qword ptr [rcx+48]
       488B09               mov      rcx, qword ptr [rcx]
       488B542430           mov      rdx, gword ptr [rsp+30H]
       395A08               cmp      dword ptr [rdx+8], ebx
       7645                 jbe      SHORT G_M43649_IG05
       8D5301               lea      edx, [rbx+1]
       895610               mov      dword ptr [rsi+16], edx
       488B16               mov      rdx, qword ptr [rsi]
       488B4110             mov      rax, qword ptr [rcx+16]
       4885C0               test     rax, rax
       750F                 jne      SHORT G_M43649_IG03
       488BCA               mov      rcx, rdx
       488D1500000000       lea      rdx, [(reloc)]
       E800000000           call     CORINFO_HELP_RUNTIMEHANDLE_CLASS

G_M43649_IG03:
       488D4C2430           lea      rcx, bword ptr [rsp+30H]
       488BD0               mov      rdx, rax
       448BC3               mov      r8d, ebx
       E800000000           call     InvariantArray`1:get_Item(int):byref:this
       488BC8               mov      rcx, rax
       488BD7               mov      rdx, rdi
       E800000000           call     CORINFO_HELP_CHECKED_ASSIGN_REF
       90                   nop      

G_M43649_IG04:
       4883C440             add      rsp, 64
       5B                   pop      rbx
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret      

G_M43649_IG05:
       488BCE               mov      rcx, rsi
       488BD7               mov      rdx, rdi
       E800000000           call     List`1:AddWithResize(ref):this
       90                   nop      

G_M43649_IG06:
       4883C440             add      rsp, 64
       5B                   pop      rbx
       5E                   pop      rsi
       5F                   pop      rdi
       C3                   ret      

; Total bytes of code 149, prolog size 19 for method List`1:Add(ref):this
```

Rather than the much reduced faster version when using `Unsafe.AsRef(in)` (this PR)
```asm
; Assembly listing for method List`1:Add(ref):this
; ...
; Lcl frame size = 0

G_M43648_IG01:
       0F1F440000           nop      

G_M43648_IG02:
       FF411C               inc      dword ptr [rcx+28]
       488B4108             mov      rax, gword ptr [rcx+8]
       448B4118             mov      r8d, dword ptr [rcx+24]
       44394008             cmp      dword ptr [rax+8], r8d
       7617                 jbe      SHORT G_M43648_IG04
       458D4801             lea      r9d, [r8+1]
       44894918             mov      dword ptr [rcx+24], r9d
       4963C8               movsxd   rcx, r8d
       488D4CC810           lea      rcx, bword ptr [rax+8*rcx+16]
       E800000000           call     CORINFO_HELP_CHECKED_ASSIGN_REF
       90                   nop      

G_M43648_IG03:
       C3                   ret      

G_M43648_IG04:
       488D0500000000       lea      rax, [(reloc)]

G_M43648_IG05:
       48FFE0               rex.jmp  rax

; Total bytes of code 55, prolog size 5 for method List`1:Add(ref):this
```
Also current for baseline
```asm
; Assembly listing for method List`1:Add(ref):this
;..
; Lcl frame size = 40

G_M43647_IG01:
       4883EC28             sub      rsp, 40
       4889542438           mov      gword ptr [rsp+38H], rdx

G_M43647_IG02:
       FF411C               inc      dword ptr [rcx+28]
       4C8B4108             mov      r8, gword ptr [rcx+8]
       8B4118               mov      eax, dword ptr [rcx+24]
       41394008             cmp      dword ptr [r8+8], eax
       761D                 jbe      SHORT G_M43647_IG04
       448D4801             lea      r9d, [rax+1]
       44894918             mov      dword ptr [rcx+24], r9d
       498BC8               mov      rcx, r8
       8BD0                 mov      edx, eax
       4C8B442438           mov      r8, gword ptr [rsp+38H]
       E800000000           call     CORINFO_HELP_ARRADDR_ST
       90                   nop      

G_M43647_IG03:
       4883C428             add      rsp, 40
       C3                   ret      

G_M43647_IG04:
       488B542438           mov      rdx, gword ptr [rsp+38H]
       488D0500000000       lea      rax, [(reloc)]

G_M43647_IG05:
       4883C428             add      rsp, 40
       48FFE0               rex.jmp  rax

; Total bytes of code 73, prolog size 9 for method List`1:Add(ref):this
```

**Comment by simplejackcoder** [comment]
@benaadams Do you know why the range check was not eliminated? I manually edited the IL after doing a private build and there is no range check. It could be that using Unsafe.AsRef may be the problem here. I am also puzzled why this PR was declined. There are only a handful of places where the arrays are private and make sense to do this, List<T> is certainly one of the best places where this makes a lot of sense.

**Pull request closed without merging**